### PR TITLE
[REVIEW] MNMG Louvain Python updates, Cython cleanup

### PR DESCRIPTION
**Pull request by rlratzel**
This adds a Python test and modules under `cugraph/dask/community` to support the MNMG Louvain C++ updates.
These changes are needed not only for exposing MG Louvain to users, but also to facilitate C++ testing.

This is still a WIP - it builds, but that's about it for now.
At a minimum, still need to:
- [ ] Improve test
- [x] Figure out Dask calls needed and update `dask/commuity/louvain.py` accordingly
- [ ] Call new 2D graph class from `louvain_wrapper.pyx` (https://github.com/rapidsai/cugraph/pull/1098)

Issue: #1052 

This also includes some cleanup: renamed the `graph_new*` files to (hopefully) better reflect their contents. The new name is `graph_primtypes*` but this can easily be changed if there's a better name that should be used.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by afender — COMMENTED** [review]
Can you elaborate on the motivation for renaming `graph` into `graph_primtypes` and why the latter is a better name?
Are we reserving  `cugraph.structure.graph` for something else?

**Comment by rlratzel** [comment]
> Can you elaborate on the motivation for renaming `graph` into `graph_primtypes` and why the latter is a better name?
> Are we reserving `cugraph.structure.graph` for something else?

I initially wanted to use "graph", but I had some issues with the graph module that cython generated colliding with the current graph.py, but I admittedly didn't push too hard to get them merged after the import errors I initially saw after that attempt.

I just took a stab at a simple rename of graph_new at that point, as well as removing the redundant graph.pxd file.  "primtypes" seemed like a decent fit based on the contents being the lower-level more primitive basic types used for graph algos.

I'm open to a better name, or even trying harder to combine the contents of graph.py with what the former graph_new cython modules were defining.  I can even back this cleanup out if we want to revisit this later.

**Comment by afender** [comment]
Typically we add the `_wrapper` suffix for `.pyx` files to avoid conflicts with `.py` files. Would it help here? I noticed there was a pyx file that didn't have the suffix. 
Ideally, we should keep the most concise name to be exposed and add prefixes/suffixes to those that don't get exposed.

**Comment by rlratzel** [comment]
> Typically we add the `_wrapper` suffix for `.pyx` files to avoid conflicts with `.py` files. Would it help here? I noticed there was a pyx file that didn't have the suffix.

I tried using that pattern as well, but I had an issue combining the code in `graph_new.pyx` and `graph_new_wrapper.pyx` into `graph_wrapper.pyx` - cython crashed, and it seemed it didn't like having a `cdef` function that was used by one of the `def` functions in the same file (my theory). I didn't investigate much and I didn't spend time refactoring it either, and instead just decided to rename for clarity and remove the redundant `graph.pxd` file. Combining the `def`s and `cdef`s this way in a single file seems like it should be legal, but the only way I could get it to build was to keep them separate (graph_new was doing this as well). So we still have 3 files, but they have hopefully better names now: `graph_primtypes.pxd`, `graph_primtypes.pyx` (cdefs), and `graph_primtypes_wrapper.pxd` (defs).

I'd like to solve this better, but I'm afraid of taking too much time away from the MG work.  I'm fine backing this out if it's too distracting now, or if we want to do it differently later.

**Comment by rlratzel** [comment]
rerun tests

**Review by ChuckHastings — APPROVED** [review]
I think this is a good start.  It would allow work to proceed in parallel while we explore things.

It would make sense for some of the things that are custom to Louvain in this PR to actually be more general purpose.  Hopefully as we get more of the cython pieces in place it will become obvious which things should migrate to more generic functions.

File: cpp/include/utilities/cython.hpp
Comment (aschaffer) [inline cpp/include/utilities/cython.hpp:19]: Consider removing dead code. If it's "experimental" code (useful for debugging, etc.) please consider guarding it between an `#ifdef - #endif` macro.

File: cpp/include/utilities/cython.hpp
Comment (aschaffer) [inline cpp/include/utilities/cython.hpp:34]: Please consider using scoped enumeration 
```
enum class weightTypeEnum : int { floatType = 0, doubleType = 1};
```
More readable, provides better name hiding (scoping), and stronger typing.

File: cpp/src/utilities/cython.cpp
Comment (aschaffer) [inline cpp/src/utilities/cython.cpp:59]: Please return-by-value,
```
return graph_container;
```
 rather than convert to r-value reference. This cast (`std::move()`) prevents the compiler to perform the very powerful return-by-value optimization (RVO), which allows elimination of copying (_copy elision_).

RVO, or copy elision allows the compiler to allocate (under a few typical circumstance, like the one here; i.e., local variable returned without any conversion) the local returned variable directly into the caller's variable that gets assigned to the result of calling the function.

**Review by aschaffer — CHANGES_REQUESTED** [review]

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp]: done in latest commit

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp]: done in latest commit

File: cpp/src/utilities/cython.cpp
Comment (rlratzel) [reply cpp/src/utilities/cython.cpp]: done in latest commit

**Review by aschaffer — APPROVED** [review]

File: python/cugraph/dask/community/louvain.py
Comment (afender) [inline python/cugraph/dask/community/louvain.py:59]: Why are we exposing symmetric DiGraph and throw errors for `Graph` or actual `DiGraph`? It is a bit awkward from a UX perspective since cuGraph has the `Graph` class for undirected graphs.

**Review by afender — APPROVED** [review]
Good progress. There's still work to do on UX, testing, and documentation

**Comment by BradReesWork** [comment]
rerun tests

**Comment by rlratzel** [comment]
rerun tests

File: python/cugraph/tests/dask/test_mg_louvain.py
Comment (BradReesWork) [reply python/cugraph/tests/dask/test_mg_louvain.py:67]: > That won't fly for the targeted MG scale but ok for small smoke tests.

CI is not for scale testing.  So not hitting the targeted MG scale in this test is fine

**Comment by BradReesWork** [comment]
rerun tests

File: cpp/include/utilities/cython.hpp
Comment (ChuckHastings) [inline cpp/include/utilities/cython.hpp:27]: Could add this in a later PR if desired.

Do we want to add GraphCSCViewFloat, GraphCSCViewDouble, GraphCOOViewFloat and GraphCOOViewDouble for completeness?  These are legacy classes, but it would allow all of the cython code to be modified to the new technique for graph allocation regardless of whether it hs been converted to the new graph objects or not.

Perhaps we defer this until we decide it would be useful.

This *might* simplify the later migration of all algorithms to MNMG.

File: cpp/include/utilities/cython.hpp
Comment (ChuckHastings) [inline cpp/include/utilities/cython.hpp:105]: Add documentation for this function.

Would be useful to understand what local_vertices, local_edges and local_offsets mean while looking at the header.

File: cpp/include/utilities/cython.hpp
Comment (ChuckHastings) [inline cpp/include/utilities/cython.hpp:67]: The destructor deletes the pointer within the union.  A reasonable thing to do.

Probably need to decide the desired life cycle behavior of this class.  If the intention is to create the graph container at the cython level and pass it by reference (probably const reference since I don't think you would want a caller to change anything) then we should delete the copy constructor, assignment operators, etc so that we won't have pointer issues.

File: cpp/include/utilities/cython.hpp
Comment (ChuckHastings) [inline cpp/include/utilities/cython.hpp:124]: Should be a const reference passed to `call_louvain`.

File: cpp/include/utilities/cython.hpp
Comment (ChuckHastings) [inline cpp/include/utilities/cython.hpp:125]: This needs to match vertex_t in the graph.  Do you want to make this a void * and cast it inside the function?

File: python/cugraph/tests/dask/test_mg_louvain.py
Comment (rlratzel) [reply python/cugraph/tests/dask/test_mg_louvain.py:73]: Agreed. I'll probably have to defer this to the next PR though.

File: python/cugraph/dask/community/louvain.py
Comment (rlratzel) [reply python/cugraph/dask/community/louvain.py:37]: Agreed, that was an oversight on my part.  I'll try to do that in the next commit to address Chuck's feedback, or may defer to the next PR.

File: python/cugraph/dask/community/louvain.py
Comment (rlratzel) [reply python/cugraph/dask/community/louvain.py:59]: Agreed, there's still some tech debt there that I may have to address in the next PR.

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp:27]: sounds good, I'll start this and see how involved it becomes. If it's holding up the PR too much I may defer it to the next PR.

**Review by BradReesWork — APPROVED** [review]

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp]: I added these and a FIXME to state how we might use them in the function when we need them.

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp:85]: Agreed. I added a copy ctor and assignment operator and deleted them.  I also added a comment in the `graph_container_t` struct describing the typical usage/lifespan of the struct, but moving forward if we need to use this differently we can support and document that as well.

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp]: done

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp]: done

File: cpp/include/utilities/cython.hpp
Comment (rlratzel) [reply cpp/include/utilities/cython.hpp]: done

**Pull request merged**
